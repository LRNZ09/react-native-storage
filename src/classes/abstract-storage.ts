abstract class AbstractStorage {
	protected constructor() {
		throw new Error(
			'[@lrnz09/react-native-storage]: Storage is not instantiable.',
		)
	}
}

export { AbstractStorage }
