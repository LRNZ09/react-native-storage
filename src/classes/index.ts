export { default as BaseStorage } from './base-storage'
export { default as Storage } from './storage'
