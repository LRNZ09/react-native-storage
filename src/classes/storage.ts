import pick from 'lodash.pick'

import BaseStorage from './base-storage'

class Storage extends BaseStorage {
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	protected static cache: Record<string, any> = {}

	static clear = async (): Promise<void> => {
		await BaseStorage.clear()

		Storage.cache = {}
	}

	static getItem = async <T>(key: string): Promise<T | null> => {
		if (key in Storage.cache) return Storage.cache[key] as T

		const value = await BaseStorage.getItem<T>(key)
		Storage.cache[key] = value

		return value
	}

	static mergeItem = async <T>(key: string, partialValue: T): Promise<void> => {
		await BaseStorage.mergeItem<T>(key, partialValue)

		const value = await BaseStorage.getItem<T>(key)
		Storage.cache[key] = value
	}

	static multiGet = async <T>(
		keys: string[],
	): Promise<[string, T | null][]> => {
		const cacheKeys = Object.keys(Storage.cache)
		const storageKeys = keys.filter((key) => !cacheKeys.includes(key))

		if (storageKeys.length > 0) {
			const entries = await BaseStorage.multiGet<T>(storageKeys)
			for (const [entryKey, entryValue] of entries)
				Storage.cache[entryKey] = entryValue
		}

		const cacheEntries = Object.entries(pick(Storage.cache, keys))

		return cacheEntries
	}

	static multiMerge = async <T>(
		partialEntries: [string, T][],
	): Promise<void> => {
		await BaseStorage.multiMerge<T>(partialEntries)

		const keys = partialEntries.map(([partialEntryKey]) => partialEntryKey)
		const entries = await BaseStorage.multiGet<T>(keys)

		for (const [entryKey, entryValue] of entries)
			Storage.cache[entryKey] = entryValue
	}

	static multiRemove = async (keys: string[]): Promise<void> => {
		await BaseStorage.multiRemove(keys)

		for (const key of keys) delete Storage.cache[key]
	}

	static multiSet = async <T>(entries: [string, T][]): Promise<void> => {
		await BaseStorage.multiSet<T>(entries)

		for (const [entryKey, entryValue] of entries)
			Storage.cache[entryKey] = entryValue
	}

	static removeItem = async (key: string): Promise<void> => {
		await BaseStorage.removeItem(key)

		delete Storage.cache[key]
	}

	static setItem = async <T>(key: string, value: T): Promise<void> => {
		await BaseStorage.setItem<T>(key, value)

		Storage.cache[key] = value
	}
}

export default Storage
