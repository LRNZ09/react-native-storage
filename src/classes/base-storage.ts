import AsyncStorage from '@react-native-async-storage/async-storage'

import {
	parseValue,
	stringifyValue,
	parseEntries,
	stringifyEntries,
} from '../utils'

import { AbstractStorage } from './abstract-storage'

// https://github.com/microsoft/TypeScript/issues/33892

class BaseStorage extends AbstractStorage /* implements StorageStatic */ {
	static clear = (): Promise<void> => AsyncStorage.clear()

	static getAllKeys = (): Promise<string[]> => AsyncStorage.getAllKeys()

	static getItem = async <T>(key: string): Promise<T | null> => {
		const rawValue = await AsyncStorage.getItem(key)
		const value = parseValue<T>(rawValue)

		return value
	}

	static mergeItem = async <T extends Record<string, any>>(
		key: string,
		value: T,
	): Promise<void> => {
		const rawValue = stringifyValue<T>(value)
		await AsyncStorage.mergeItem(key, rawValue)
	}

	static multiGet = async <T>(
		keys: string[],
	): Promise<[string, T | null][]> => {
		const rawEntries = await AsyncStorage.multiGet(keys)
		const entries = parseEntries<T>(rawEntries)

		return entries
	}

	static multiMerge = async <T extends Record<string, any>>(
		entries: [string, T][],
	): Promise<void> => {
		const rawEntries = stringifyEntries<T>(entries)
		await AsyncStorage.multiMerge(rawEntries)
	}

	static multiRemove = (keys: string[]): Promise<void> =>
		AsyncStorage.multiRemove(keys)

	static multiSet = async <T>(entries: [string, T][]): Promise<void> => {
		const rawEntries = stringifyEntries<T>(entries)
		await AsyncStorage.multiSet(rawEntries)
	}

	static removeItem = (key: string): Promise<void> =>
		AsyncStorage.removeItem(key)

	static setItem = <T>(key: string, value: T): Promise<void> =>
		AsyncStorage.setItem(key, stringifyValue<T>(value))
}

export { BaseStorage }
