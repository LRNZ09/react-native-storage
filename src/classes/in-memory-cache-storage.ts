import pick from 'lodash.pick'

import { BaseStorage } from './base-storage'

class InMemoryCacheStorage extends BaseStorage {
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	protected static cache: Record<string, any> = {}

	static clear = async (): Promise<void> => {
		await BaseStorage.clear()

		InMemoryCacheStorage.cache = {}
	}

	static getItem = async <T>(key: string): Promise<T | null> => {
		if (key in InMemoryCacheStorage.cache)
			return InMemoryCacheStorage.cache[key] as T

		const value = await BaseStorage.getItem<T>(key)
		InMemoryCacheStorage.cache[key] = value

		return value
	}

	static mergeItem = async <T>(key: string, partialValue: T): Promise<void> => {
		await BaseStorage.mergeItem<T>(key, partialValue)

		const value = await BaseStorage.getItem<T>(key)
		InMemoryCacheStorage.cache[key] = value
	}

	static multiGet = async <T>(
		keys: string[],
	): Promise<[string, T | null][]> => {
		const cacheKeys = Object.keys(InMemoryCacheStorage.cache)
		const storageKeys = keys.filter((key) => !cacheKeys.includes(key))

		if (storageKeys.length > 0) {
			const entries = await BaseStorage.multiGet<T>(storageKeys)
			for (const [entryKey, entryValue] of entries)
				InMemoryCacheStorage.cache[entryKey] = entryValue
		}

		const cacheEntries = Object.entries(pick(InMemoryCacheStorage.cache, keys))

		return cacheEntries
	}

	static multiMerge = async <T>(
		partialEntries: [string, T][],
	): Promise<void> => {
		await BaseStorage.multiMerge<T>(partialEntries)

		const keys = partialEntries.map(([partialEntryKey]) => partialEntryKey)
		const entries = await BaseStorage.multiGet<T>(keys)

		for (const [entryKey, entryValue] of entries)
			InMemoryCacheStorage.cache[entryKey] = entryValue
	}

	static multiRemove = async (keys: string[]): Promise<void> => {
		await BaseStorage.multiRemove(keys)

		for (const key of keys) delete InMemoryCacheStorage.cache[key]
	}

	static multiSet = async <T>(entries: [string, T][]): Promise<void> => {
		await BaseStorage.multiSet<T>(entries)

		for (const [entryKey, entryValue] of entries)
			InMemoryCacheStorage.cache[entryKey] = entryValue
	}

	static removeItem = async (key: string): Promise<void> => {
		await BaseStorage.removeItem(key)

		delete InMemoryCacheStorage.cache[key]
	}

	static setItem = async <T>(key: string, value: T): Promise<void> => {
		await BaseStorage.setItem<T>(key, value)

		InMemoryCacheStorage.cache[key] = value
	}
}

export { InMemoryCacheStorage }
