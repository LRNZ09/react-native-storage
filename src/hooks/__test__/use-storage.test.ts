import { renderHook } from '@testing-library/react-hooks'

import { useStorage } from '..'
import { BaseStorage } from '../../classes'

const ITEM_KEY = '@foobar'

describe('useStorage', () => {
	beforeEach(async () => {
		await BaseStorage.setItem(ITEM_KEY, { foo: 1 })
	})

	afterEach(async () => {
		await BaseStorage.clear()
	})

	test('should return functions for a given item key', () => {
		const { result } = renderHook(() => useStorage(ITEM_KEY))

		expect(result.current).toMatchSnapshot()
	})

	test('should get item from hook function', async () => {
		const { result } = renderHook(() => useStorage(ITEM_KEY))

		expect(await result.current.getItem()).toEqual({ foo: 1 })
	})

	test('should set item from hook function', async () => {
		const { result } = renderHook(() => useStorage<{ bar: number }>(ITEM_KEY))

		expect(await result.current.setItem({ bar: 42 })).toBeUndefined()
		expect(await result.current.getItem()).toEqual({ bar: 42 })
	})

	test('should merge item from hook function', async () => {
		const { result } = renderHook(() =>
			useStorage<{ foo: number; bar: number }>(ITEM_KEY),
		)

		expect(await result.current.mergeItem({ bar: 42 })).toBeUndefined()
		expect(await result.current.getItem()).toEqual({ foo: 1, bar: 42 })
	})

	test('should remove item from hook function', async () => {
		const { result } = renderHook(() => useStorage(ITEM_KEY))

		expect(await result.current.removeItem()).toBeUndefined()
		expect(await result.current.getItem()).toBeNull()
	})
})
