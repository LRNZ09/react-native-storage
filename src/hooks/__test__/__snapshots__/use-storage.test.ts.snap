// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useStorage should return functions for a given item key 1`] = `
Object {
  "getItem": [Function],
  "mergeItem": [Function],
  "removeItem": [Function],
  "setItem": [Function],
}
`;
